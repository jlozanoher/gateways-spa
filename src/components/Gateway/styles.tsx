import styled from "styled-components";
import * as Antd from "antd";

export const Header = styled.div`
  display: flex;
  align-items: center;
  padding: 8px 16px;
`;

export const Grow = styled.div`
  flex-grow: 1;
`;

export const Modal = styled(Antd.Modal)`
  padding: 40px 24px;
`;
