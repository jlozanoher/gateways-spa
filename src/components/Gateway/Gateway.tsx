import { DeleteOutlined } from "@ant-design/icons";
import { Button, Typography } from "antd";
import React, { useContext } from "react";
import { useDrop } from "react-dnd";
import { PeripheralContext } from "../../App";
import { useGatewayDelete } from "../../contollers/gateway.controller";
import { usePeripheralUpdate } from "../../contollers/peripheral.controller";
import { dotsString } from "../../lib/utils";
import { GatewayModel } from "../../types/gateway.model";
import { ItemTypes } from "../../types/ItemsTypes";
import { PeripheralList } from "../Peripheral/PeripheralList";
import * as S from "../styles";
import { GatewayContext } from "./Gateways";

const { Title, Text } = Typography;

interface Props {
  gateway: GatewayModel;
}

export const Gateway = (props: Props) => {
  const { gateway } = props;

  const { fetchGateways = () => {} } = useContext(GatewayContext);
  const { peripherals, fetchPeripherals = () => {} } =
    useContext(PeripheralContext);

  const { handleUpdatePeripheral } = usePeripheralUpdate({
    onSuccess: () => {
      fetchPeripherals();
    },
  });

  const [collected, drop] = useDrop(
    () => ({
      accept: ItemTypes.PERIPHERAL,
      drop: (item: any, monitor: any) => {
        handleUpdatePeripheral({ ...item, gateway: gateway._id });
      },
      collect: (monitor) => {
        return monitor.getItem();
      },
    }),
    []
  );

  const { handleDeleteGateway } = useGatewayDelete({
    onSuccess: () => {
      fetchGateways();
      fetchPeripherals();
    },
  });

  const onDelete = () => {
    handleDeleteGateway(gateway._id);
  };

  const attachedPeripherals = peripherals?.filter(
    (p) => p.gateway === gateway._id
  );

  return (
    <div ref={drop}>
      <S.Card>
        <S.Box>
          <S.Box>
            <Button icon={<DeleteOutlined />} danger onClick={onDelete} />
          </S.Box>
          <S.Box style={{ flexDirection: "column" }}>
            <Title level={5}>{gateway.name}</Title>
            <Text type="secondary">{dotsString(gateway._id)}</Text>
            <div>{gateway.serialNumber}</div>
            <div>{gateway.ipv4Address}</div>
          </S.Box>
          <S.Box style={{ flexDirection: "column", flexGrow: 1 }}>
            <Title level={5}>{"Peripherals"}</Title>
            <PeripheralList peripherals={attachedPeripherals || []} />
          </S.Box>
        </S.Box>
      </S.Card>
    </div>
  );
};
