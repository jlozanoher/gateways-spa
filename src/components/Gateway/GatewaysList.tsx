import React from "react";
import * as S from "../styles";

import { GatewayModel } from "../../types/gateway.model";
import { Gateway } from "./Gateway";

interface Props {
  gateways: GatewayModel[];
}

export const GatewaysList = (props: Props) => {
  const { gateways } = props;

  return (
    <S.DivH>
      {gateways.map((gateway) => (
        <Gateway key={gateway._id} gateway={gateway} />
      ))}
    </S.DivH>
  );
};
