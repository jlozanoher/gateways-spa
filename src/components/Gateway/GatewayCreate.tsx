import { Button, Form, Input } from "antd";
import React, { useContext, useEffect } from "react";
import {
  useGatewayCreate,
  useGatewayUpdate,
} from "../../contollers/gateway.controller";
import { GatewayModel } from "../../types/gateway.model";
import { GatewayContext } from "./Gateways";
import * as S from "../styles";

const layout = {
  labelCol: { span: 6 },
  wrapperCol: { span: 16 },
};
const tailLayout = {
  wrapperCol: { offset: 8, span: 16 },
};

interface Props {
  gateway?: GatewayModel; // For updating instead of creating
}

export const GatewayCreate = (props: Props) => {
  const { gateway } = props;

  const { fetchGateways = () => {} } = useContext(GatewayContext);

  const { handleCreateGateway } = useGatewayCreate({
    onSuccess: () => {
      fetchGateways();
    },
    onError: (err) => {
      if (err?.name?.code === 11000 && err?.name?.keyPattern?.serialNumber)
        form.setFields([
          {
            name: "serialNumber",
            errors: ["Duplicity"],
          },
        ]);
    },
  });
  const { handleUpdateGateway } = useGatewayUpdate({
    onSuccess: () => {
      fetchGateways();
    },
  });

  const [form] = Form.useForm();

  useEffect(() => {
    if (gateway) {
      form.setFieldsValue(gateway);
    }
  }, [gateway]);

  const onFinish = (values: any) => {
    if (gateway?._id) {
      handleUpdateGateway({ ...values, _id: gateway._id });
    } else {
      handleCreateGateway(values);
    }
  };

  const onClear = () => {
    form.resetFields();
  };

  return (
    <div>
      <Form {...layout} form={form} onFinish={onFinish}>
        <Form.Item name={"name"} label={"Name"} rules={[{ required: true }]}>
          <Input />
        </Form.Item>
        <Form.Item
          name={"serialNumber"}
          label={"Serial number"}
          rules={[{ required: true }]}
        >
          <Input />
        </Form.Item>
        <Form.Item
          name={"ipv4Address"}
          label={"IPv4"}
          rules={[
            { required: true },
            () => ({
              validator(rule, value) {
                if (
                  !/^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/.test(
                    value
                  )
                ) {
                  return Promise.reject("Invalid IPv4 address!");
                }
                return Promise.resolve();
              },
            }),
          ]}
        >
          <Input />
        </Form.Item>
        <Form.Item {...tailLayout}>
          <S.Box>
            <Button type="primary" htmlType="submit">
              {gateway ? "Update Gateway" : "Create Gateway"}
            </Button>
            <Button onClick={onClear}>Clear</Button>
          </S.Box>
        </Form.Item>
      </Form>
    </div>
  );
};
