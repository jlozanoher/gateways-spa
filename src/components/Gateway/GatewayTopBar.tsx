import { PlusOutlined, WarningOutlined } from "@ant-design/icons";
import { Button, Typography } from "antd";
import React, { useContext } from "react";
import { GatewayContext } from "./Gateways";
import * as S from "../styles";

const { Title, Text } = Typography;

export const GatewayTopBar = () => {
  const { setShowModal = () => {} } = useContext(GatewayContext);

  const onClick = () => {
    setShowModal(true);
  };

  return (
    <S.Header>
      <Title level={3}>Gateways</Title>
      <S.Grow>
        <Text type="warning">
          <WarningOutlined />
          Drag the peripherals into or out a Gateway
          <WarningOutlined />
        </Text>
      </S.Grow>
      <Button type="primary" icon={<PlusOutlined />} onClick={onClick}>
        Add Gateway
      </Button>
    </S.Header>
  );
};
