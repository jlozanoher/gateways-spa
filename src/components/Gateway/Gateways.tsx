import { Button } from "antd";
import axios from "axios";
import React, { Dispatch, SetStateAction, useEffect, useState } from "react";
import { ROUTES } from "../../lib/routes";
import { GatewayModel } from "../../types/gateway.model";
import * as S from "../styles";
import { GatewayCreate } from "./GatewayCreate";
import { GatewaysList } from "./GatewaysList";
import { GatewayTopBar } from "./GatewayTopBar";

export interface IGatewayContext {
  fetchGateways?: () => void;
  setShowModal?: Dispatch<SetStateAction<boolean>>;
  showModal?: boolean;
  gatewayForEditon?: GatewayModel;
  setGatewayForEdition?: Dispatch<SetStateAction<GatewayModel | undefined>>;
}

export const GatewayContext = React.createContext<IGatewayContext>({});

export const Gateways = () => {
  const [gateways, setGateways] = useState<GatewayModel[]>([]);
  const [gatewayForEditon, setGatewayForEdition] = useState<
    GatewayModel | undefined
  >(undefined);
  const [showModal, setShowModal] = useState(false);

  const fetchGateways = () => {
    axios
      .get(`${process.env.REACT_APP_API_URL}${ROUTES.gateways}`)
      .then((res) => {
        setGateways(res.data);
      });
  };

  useEffect(() => {
    fetchGateways();
  }, []);

  const handleModalClose = () => {
    setShowModal(false);
    setGatewayForEdition(undefined);
  };

  return (
    <GatewayContext.Provider
      value={{
        fetchGateways,
        setShowModal,
        showModal,
        gatewayForEditon,
        setGatewayForEdition,
      }}
    >
      <GatewayTopBar />
      <GatewaysList gateways={gateways} />
      <S.Modal
        title={gatewayForEditon ? "Update Gateway" : "Create Gateway"}
        maskClosable
        visible={showModal}
        onCancel={handleModalClose}
        footer={[
          <Button type="primary" onClick={handleModalClose}>
            Cancel
          </Button>,
        ]}
      >
        <GatewayCreate gateway={gatewayForEditon} />
      </S.Modal>
    </GatewayContext.Provider>
  );
};
