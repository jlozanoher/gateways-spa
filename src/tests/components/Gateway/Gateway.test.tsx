import React from "react";
import renderer from "react-test-renderer";
import { Gateway } from "../../../components/Gateway/Gateway";
import { GatewayModel } from "../../../types/gateway.model";
import { WithDnDContext } from "../../test_utils";

const gateway: GatewayModel = {
  _id: "123sda1",
  name: "Gate 1",
  ipv4Address: "1.1.1.1",
  serialNumber: "111",
};

test("Gateway should match snap", () => {
  const tree = renderer
    .create(
      <WithDnDContext>
        <Gateway gateway={gateway} />
      </WithDnDContext>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
