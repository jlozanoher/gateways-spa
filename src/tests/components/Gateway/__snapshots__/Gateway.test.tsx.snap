// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gateway should match snap 1`] = `
<div>
  <div
    className="ant-card ant-card-bordered sc-jRQBWg kCIKTS"
  >
    <div
      className="ant-card-body"
      style={Object {}}
    >
      <div
        className="sc-hKwDye dbURyo"
      >
        <div
          className="sc-eCImPb gyozil"
        >
          <button
            className="ant-btn ant-btn-icon-only ant-btn-dangerous"
            onClick={[Function]}
            type="button"
          >
            <span
              aria-label="delete"
              className="anticon anticon-delete"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="delete"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                />
              </svg>
            </span>
          </button>
          <button
            className="ant-btn ant-btn-icon-only"
            onClick={[Function]}
            type="button"
          >
            <span
              aria-label="edit"
              className="anticon anticon-edit"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="edit"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M257.7 752c2 0 4-.2 6-.5L431.9 722c2-.4 3.9-1.3 5.3-2.8l423.9-423.9a9.96 9.96 0 000-14.1L694.9 114.9c-1.9-1.9-4.4-2.9-7.1-2.9s-5.2 1-7.1 2.9L256.8 538.8c-1.5 1.5-2.4 3.3-2.8 5.3l-29.5 168.2a33.5 33.5 0 009.4 29.8c6.6 6.4 14.9 9.9 23.8 9.9zm67.4-174.4L687.8 215l73.3 73.3-362.7 362.6-88.9 15.7 15.6-89zM880 836H144c-17.7 0-32 14.3-32 32v36c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-36c0-17.7-14.3-32-32-32z"
                />
              </svg>
            </span>
          </button>
        </div>
        <div
          className="sc-eCImPb gyozil"
        >
          <h5
            className="ant-typography"
            onClick={null}
            style={
              Object {
                "WebkitLineClamp": undefined,
              }
            }
          >
            Gate 1
          </h5>
          <span
            className="ant-typography ant-typography-secondary"
            onClick={null}
            style={
              Object {
                "WebkitLineClamp": undefined,
              }
            }
          >
            123s....sda1
          </span>
          <div>
            111
          </div>
          <div>
            1.1.1.1
          </div>
        </div>
        <div
          className="sc-hKwDye dbURyo"
          style={
            Object {
              "flexDirection": "column",
              "flexGrow": 1,
            }
          }
        >
          <h5
            className="ant-typography"
            onClick={null}
            style={
              Object {
                "WebkitLineClamp": undefined,
              }
            }
          >
            Peripherals
          </h5>
          <div
            className="ant-list ant-list-split ant-list-grid"
          >
            <div
              className="ant-spin-nested-loading"
            >
              <div
                className="ant-spin-container"
              >
                <div
                  className="ant-list-empty-text"
                >
                  <div
                    className="ant-empty ant-empty-normal"
                  >
                    <div
                      className="ant-empty-image"
                    >
                      <svg
                        className="ant-empty-img-simple"
                        height="41"
                        viewBox="0 0 64 41"
                        width="64"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <g
                          fill="none"
                          fillRule="evenodd"
                          transform="translate(0 1)"
                        >
                          <ellipse
                            className="ant-empty-img-simple-ellipse"
                            cx="32"
                            cy="33"
                            rx="32"
                            ry="7"
                          />
                          <g
                            className="ant-empty-img-simple-g"
                            fillRule="nonzero"
                          >
                            <path
                              d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                            />
                            <path
                              className="ant-empty-img-simple-path"
                              d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                            />
                          </g>
                        </g>
                      </svg>
                    </div>
                    <div
                      className="ant-empty-description"
                    >
                      No Data
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
